modules/gdscript/gdscript_editor.cpp

```diff
@@ -1932,6 +1932,11 @@ static bool _guess_identifier_type(GDScriptParser::CompletionContext &p_context,
 				default:
 					break;
 			}
+
+			if (GDScriptParser::DataType::TypeSource::INFERRED <= id_type.type_source) {
```

Comment:
```suggestion
			if (!id_type.has_no_type()) {
```

or

```suggestion
			if (id_type.type_source != GDScriptParser::DataType::UNDETECTED) {
```

But I think this is a too early return, which will cause the checks below to be ignored, and autocompletion will always use the weak type. I think it would be more correct to move this below to use the hard type only if there is no weak type.

Reply:
OK, I use `type_source != GDScriptParser::DataType::UNDETECTED`.

> But I think this is a too early return

So, I changed the code to use the weak type only if the result of `_guess_expression_type` is variant.